File: common/http/test/vault-server.pl
#!/usr/bin/perl
use strict;
use warnings;

use HTTP::Daemon::SSL;
use HTTP::Status;
use HTTP::Request;
use JSON;

my $pem_file = '/tmp/vault.pem';

system("openssl req -new -x509 -newkey rsa:2048 -nodes -keyout $pem_file -out $pem_file -days 365 -subj /CN=localhost");

# Création du serveur HTTPS sur le port 4443
my $d = HTTP::Daemon::SSL->new(
    LocalPort => 4443,
    SSL_cert_file => $pem_file,
    SSL_key_file  => $pem_file,
) || die "Impossible de démarrer le serveur HTTPS: $!";

print "Serveur HTTPS démarré sur ", $d->url, "\n";

# Boucle principale pour accepter les connexions
while (my $client_conn = $d->accept) {
    while (my $request = $client_conn->get_request) {
        # Gérer les requêtes POST sur le chemin /v1/auth/approle/login
        if ($request->method eq 'POST' && $request->uri->path eq '/v1/auth/approle/login') {

            # Récupérer les données du corps de la requête POST
            my $content = $request->content;

            # Parse du contenu JSON de la requête (si c'est du JSON)
            my $json_data = eval { decode_json($content) };
            if ($@) {
                # Si erreur dans le JSON, renvoyer une erreur 400 (Bad Request)
                $client_conn->send_response(400, 'Bad Request', undef, 'Invalid JSON');
            } else {
                # Simuler une réponse JSON pour le login
		my $response_data = {
		  request_id => "aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee",
		  lease_id => "",
		  renewable => JSON::false,
		  lease_duration => 0,
		  data => undef,
		  wrap_info => undef,
		  warnings => undef,
		  auth => {
		    client_token => "hvs.key that does not exist",
		    accessor => "A0A0A0A0A0A0A0A0A0A0A0A0",
		    policies => ["default", "john-doe"],
		    token_policies => ["default", "john-doe"],
		    metadata => {
		      role_name => "john-doe"
		    },
		    lease_duration => 2764800,
		    renewable => JSON::true,
		    entity_id => "bbbbbbbb-bbbb-cccc-dddd-ffffffffffff",
		    token_type => "service",
		    orphan => JSON::true,
		    mfa_requirement => undef,
		    num_uses => 0
		  },
		  mount_type => ""
		};

                # Envoyer une réponse JSON
                $client_conn->send_response(200, 'OK', undef, encode_json($response_data));
            }
        } else {
            # Gérer les autres chemins (404 Not Found)
            $client_conn->send_error(RC_NOT_FOUND, "Path not found");
        }
    }
    $client_conn->close;
    undef($client_conn);
}

File: common/http/test/vault-server.py
#!/usr/bin/env python3
"""
Minimal HTTPS server simulating a HashiCorp Vault approle login endpoint.
Requires only standard library modules (ssl, http.server, json, subprocess).
All diagnostic output goes to stdout so the test framework can capture it.
"""
print("vault-server.py: starting", flush=True)

import json
import ssl
import subprocess
import sys
from http.server import BaseHTTPRequestHandler, HTTPServer


def log(msg):
    print(msg, flush=True)


class VaultHandler(BaseHTTPRequestHandler):
    def log_message(self, fmt, *args):
        pass  # silence default access log

    def do_POST(self):
        if self.path == "/v1/auth/approle/login":
            length = int(self.headers.get("Content-Length", 0))
            body = self.rfile.read(length)
            try:
                json.loads(body)
            except (json.JSONDecodeError, ValueError):
                self.send_response(400)
                self.end_headers()
                self.wfile.write(b"Invalid JSON")
                return

            payload = {
                "request_id": "aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee",
                "lease_id": "",
                "renewable": False,
                "lease_duration": 0,
                "data": None,
                "wrap_info": None,
                "warnings": None,
                "auth": {
                    "client_token": "hvs.key that does not exist",
                    "accessor": "A0A0A0A0A0A0A0A0A0A0A0A0",
                    "policies": ["default", "john-doe"],
                    "token_policies": ["default", "john-doe"],
                    "metadata": {"role_name": "john-doe"},
                    "lease_duration": 2764800,
                    "renewable": True,
                    "entity_id": "bbbbbbbb-bbbb-cccc-dddd-ffffffffffff",
                    "token_type": "service",
                    "orphan": True,
                    "mfa_requirement": None,
                    "num_uses": 0,
                },
                "mount_type": "",
            }
            data = json.dumps(payload).encode()
            self.send_response(200)
            self.send_header("Content-Type", "application/json")
            self.send_header("Content-Length", str(len(data)))
            self.end_headers()
            self.wfile.write(data)
        else:
            self.send_response(404)
            self.end_headers()
            self.wfile.write(b"Not found")


try:
    key_file = "/tmp/vault-key.pem"
    cert_file = "/tmp/vault-cert.pem"

    log("vault-server.py: generating certificate")
    # Use -noenc (OpenSSL 3.x) with a fallback to -nodes (OpenSSL 1.x).
    for noenc_flag in ("-noenc", "-nodes"):
        result = subprocess.run(
            [
                "openssl", "req", "-new", "-x509", "-newkey", "rsa:2048",
                noenc_flag, "-keyout", key_file, "-out", cert_file,
                "-days", "365", "-subj", "/CN=localhost",
            ],
            capture_output=True,
        )
        log(f"vault-server.py: openssl {noenc_flag} returned {result.returncode}")
        if result.returncode == 0:
            break
        log(f"openssl stderr: {result.stderr.decode()}")
    if result.returncode != 0:
        log("vault-server.py: openssl failed with both -noenc and -nodes")
        sys.exit(1)

    log("vault-server.py: creating SSL context")
    ctx = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)

    log("vault-server.py: loading cert chain")
    ctx.load_cert_chain(certfile=cert_file, keyfile=key_file)

    log("vault-server.py: binding on 0.0.0.0:4443")
    server = HTTPServer(("0.0.0.0", 4443), VaultHandler)

    log("vault-server.py: wrapping socket with SSL")
    server.socket = ctx.wrap_socket(server.socket, server_side=True)

    log("vault-server.py: ready, serving forever")
    server.serve_forever()

except Exception as e:
    log(f"vault-server.py: ERROR: {type(e).__name__}: {e}")
    sys.exit(1)
